Java/benchmark/src/main/java/com/amazon/randomcutforest/ForestUpdateExecutorBenchmark.java

```diff
@@ -30,12 +29,18 @@
 import org.openjdk.jmh.annotations.Warmup;
 import org.openjdk.jmh.infra.Blackhole;
 
-import com.amazon.randomcutforest.sampler.SimpleStreamSamplerV2;
-import com.amazon.randomcutforest.store.PointStore;
+import com.amazon.randomcutforest.executor.AbstractForestUpdateExecutor;
+import com.amazon.randomcutforest.executor.IStateCoordinator;
+import com.amazon.randomcutforest.executor.ParallelForestUpdateExecutor;
+import com.amazon.randomcutforest.executor.PassThroughCoordinator;
+import com.amazon.randomcutforest.executor.PointStoreCoordinator;
+import com.amazon.randomcutforest.executor.SamplerPlusTree;
+import com.amazon.randomcutforest.executor.SequentialForestUpdateExecutor;
+import com.amazon.randomcutforest.sampler.SimpleStreamSampler;
+import com.amazon.randomcutforest.store.PointStoreDouble;
 import com.amazon.randomcutforest.testutils.NormalMixtureTestData;
-import com.amazon.randomcutforest.tree.CompactRandomCutTree;
+import com.amazon.randomcutforest.tree.CompactRandomCutTreeDouble;
```

Comment:
I've seen this pattern where a `Class` is defined with generics and then `ClassFloat` and `ClassDouble` are defined as `Class<float>` and `Class<double>`. Though, there are some instances like in `IndexManager` where the float type is not exactly the same as generic: `IndexManager<float[], float[]>`. Nonetheless, Is there a reason why the user can't do this? That is, instead of having three versions of similar functionality why not have one generic one and the user just defines a `RandomCutForest<Double>` in their code? Or do these classes have float type-specific operations?

Reply:
This is due to a Java limitation: Java doesn't support classes that are generic over unboxed primitive types. You can have `Class<Double>` with a capital `D` corresponding to the boxed type, but then every `Double` in the class is a pointer to a `double` value, and we take a hit for all the dereferencing. In the places where we created `ClassDouble` and `ClassFloat` versions of classes, it should either be because there was some math we wanted to do with primitive floats and doubles, or else because the class depends on another class that does math with primitive floats and doubles. That said, there was a lot of refactoring that happened after these classes were first created, and it would be good to review all of them to make sure that we really need `Double` and `Float` versions still.

Reply:
Indeed, this is a plus for Rust and would not be needed there. The pattern is over primarily over [BoundingBox, Tree, PointStore, BoxCache] and then it spills over to the mapper classes. PointStore would need this precision setup (checking equality etc.). Now if the PointStore converts these to some standard (say double[]) to pass to the tree (since the tree coordinates the bounding boxes)  then there is multiple conversions of precision which seems like a bad idea (plus the precision of the random cuts becomes an issue in these conversions). That kind of indicates a split for the trees is needed. The boxes (basic class and the cached) are interesting -- it may be possible to refactor them (i.e., go via PointStore). But the Visitors were using the double precision anyways and we chose to keep the Visitors as intact as possible. So the refactoring would likely involve the Visitors. Right now the precision set up is internal -- the input or the visitors see double[]. Conversions are done exactly at two places (once in input, at PointStore and once in egress, while exposing to the visitor in CompactNodeView).

Reply:
(Thanks for the explanation. FWIW, I wasn't trying to flex any Rust muscles nor did I know that this was a Java language limitation.)

@jotok I can help look for the places you mentioned where the two versions might not be needed.

@sudiptoguha It's best to avoid multiple conversations for floating point accuracy and performance reasons, so I'm not advocating in any way to remove the `Float` vs. `Double` distinction. But having to handle both types explicitly is a bit of a scaling problem that, apparently, cannot be avoided.

Reply:
I think that in Java we can support 32-bit and 64-bit floating point numbers, and say that if you want more control than that then you'll have to use a different implementation (i.e., Rust). So this bit of ugliness should at least not get any worse than it currently is.